programs/ssl/ssl_client2.c

```diff
@@ -3059,7 +3064,15 @@ int main( int argc, char *argv[] )
 #endif
 
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
+
     mbedtls_psa_crypto_free( );
+    const char* message = mbedtls_test_helper_is_psa_leaking();
```

Comment:
When I moved `mbedtls_test_helper_is_psa_leaking()` before `mbedtls_psa_crypto_free()` it turned out that there is a problem but on server app only. It reports `A volatile slot has not been closed properly`. Checked also that there are 2 volatile slots open.
This check is done after calling `psa_destroy_key( key_slot ); psa_destroy_key( key_slot2 );` so I assume that this should do the job. Further investigation is required.

Reply:
Hmm, on one hand hurray, because this is the kind of things we wanted to catch, but on the other hand, that's a new thing to investigate... I have trouble reproducing locally, can you tell me how you're getting this error?

Reply:
I found this by running opaque test like:
`przemek@przemek-VirtualBox:~/src/mbedtls/tests$ ./ssl-opt.sh -f "Opaque key for client/server authentication"`

```
przemek@przemek-VirtualBox:~/src/mbedtls/tests$ cat o-srv-1.log 
# Opaque key for client/server authentication
../programs/ssl/ssl_server2 server_addr=127.0.0.1 server_port=17814 allow_sha1=1 auth_mode=required key_opaque=1 crt_file=data_files/server5.crt              key_file=data_files/server5.key

  . Seeding the random number generator... ok
  . Loading the CA root certificate ... ok (0 skipped)
  . Loading the server cert. and key... ok (key types: Opaque - invalid PK)
  . Setting up the SSL/TLS structure... ok
  . Bind on tcp://127.0.0.1:17814/ ... ok
  . Waiting for a remote connection ... ok
  . Performing the SSL/TLS handshake... ok
    [ Protocol is TLSv1.2 ]
    [ Ciphersuite is TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256 ]
    [ Record expansion is 21 ]
    [ Maximum incoming record payload length is 16384 ]
    [ Maximum outgoing record payload length is 16384 ]
  . Verifying peer X.509 certificate... ok
  . Peer certificate information    ...
      cert. version     : 3
      serial number     : 09
      issuer name       : C=NL, O=PolarSSL, CN=Polarssl Test EC CA
      subject name      : C=NL, O=PolarSSL, CN=localhost
      issued  on        : 2013-09-24 15:52:04
      expires on        : 2023-09-22 15:52:04
      signed using      : ECDSA with SHA256
      EC key size       : 256 bits
      basic constraints : CA=false

  < Read from client: 34 bytes read

GET / HTTP/1.0
Extra-header: 


  > Write to client: 158 bytes written in 1 fragments

HTTP/1.0 200 OK
Content-Type: text/html

<h2>mbed TLS Test Server</h2>
<p>Successful connection using: TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256</p>

  . Closing the connection... done
  . Waiting for a remote connection ... interrupted by SIGTERM (in net_accept())
  . Cleaning up...Leak Stats 2 0 0 0 0 <--- leak stats (2 volatile slots active)
PSA memory leak detected: A volatile slot has not been closed properly. <---- error message
 done.
```

Then I tried to run both apps manually and adding more traces on different stages of test execution.
It seems that this is not related to psa key slot, but `mbedtls_ssl_ticket_setup()`:

```
przemek@przemek-VirtualBox:~/src/mbedtls/programs/ssl$ ./ssl_server2 server_addr=127.0.0.1 server_port=10788 auth_mode=required key_opaque=1 crt_file=../../tests/data_files/server5.crt key_file=../../tests/data_files/server5.key

  . Seeding the random number generator... ok
  . Loading the CA root certificate ... ok (0 skipped)
  . Loading the server cert. and key...
====> Before mbedtls_pk_wrap_as_opaque() 
Leak Stats 0 0 0 0 0 

====> After mbedtls_pk_wrap_as_opaque() 
Leak Stats 1 0 0 0 0 
 ok (key types: Opaque - invalid PK)
  . Setting up the SSL/TLS structure...
====> Before mbedtls_ssl_ticket_setup() 
Leak Stats 1 0 0 0 0 

====> After mbedtls_ssl_ticket_setup() 
Leak Stats 3 0 0 0 0 
 ok
  . Bind on tcp://127.0.0.1:10788/ ... ok
  . Waiting for a remote connection ... ok
  . Performing the SSL/TLS handshake... ok
    [ Protocol is TLSv1.2 ]
    [ Ciphersuite is TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256 ]
    [ Record expansion is 21 ]
    [ Maximum incoming record payload length is 16384 ]
    [ Maximum outgoing record payload length is 16384 ]
  . Verifying peer X.509 certificate... ok
  . Peer certificate information    ...
      cert. version     : 3
      serial number     : 09
      issuer name       : C=NL, O=PolarSSL, CN=Polarssl Test EC CA
      subject name      : C=NL, O=PolarSSL, CN=localhost
      issued  on        : 2013-09-24 15:52:04
      expires on        : 2023-09-22 15:52:04
      signed using      : ECDSA with SHA256
      EC key size       : 256 bits
      basic constraints : CA=false

  < Read from client: 34 bytes read

GET / HTTP/1.0
Extra-header: 


  > Write to client: 158 bytes written in 1 fragments

HTTP/1.0 200 OK
Content-Type: text/html

<h2>mbed TLS Test Server</h2>
<p>Successful connection using: TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256</p>

  . Closing the connection... done
  . Cleaning up...
====> Before psa_destroy_key() 
Leak Stats 3 0 0 0 0 

====> After psa_destroy_key() 
Leak Stats 2 0 0 0 0 
Leak Stats 2 0 0 0 0 
PSA memory leak detected: A volatile slot has not been closed properly.
 done.

```

@mpg From the above log we can see that slot counter increases on `mbedtls_pk_wrap_as_opaque` (+1) and `mbedtls_ssl_ticket_setup` (+3) and decreases on `psa_destroy_key` (-1).

As you requested I moved memory leak check before `mbedtls_psa_crypto_free( );` (`rng_free()` also calls `mbedtls_psa_crypto_free( ); - is this intended?`). Now I see two slots taken by ssl ticket. They are released after calling `rng_free()`.

Reply:
`mbedtls_ssl_ticket_free` should have freed the keys created by `mbedtls_ssl_ticket_setup`, I guess that's where the bug is.

> rng_free() also calls mbedtls_psa_crypto_free( ); - is this intended?

I wrote it this way so that it frees all resources in applications that otherwise don't use PSA crypto, which happens in a build with `MBEDTLS_TEST_USE_PSA_CRYPTO_RNG` but without `MBEDTLS_USE_PSA_CRYPTO`. As a consequence, in an aplication that calls `mbedtls_test_helper_is_psa_leaking` with the expectation that all keys are destroyed, it must do so before calling `rng_free`. Can you please add a note in the documentation of `rng_free` to indicate that it calls `mbedtls_psa_crypto_free`, and as a consequence if you want to call `mbedtls_test_helper_is_psa_leaking` you must do this first?

Reply:
> mbedtls_ssl_ticket_free should have freed the keys created by mbedtls_ssl_ticket_setup, I guess that's where the bug is.

If I switch order of the following code blocks as follows, then no memory leak is detected and `mbedtls_test_helper_is_psa_leaking` is called before `rng_free`, but not sure if reversing order is ok:

```
#if defined(MBEDTLS_SSL_CACHE_C)
    mbedtls_ssl_cache_free( &cache );
#endif
#if defined(MBEDTLS_SSL_SESSION_TICKETS)
    mbedtls_ssl_ticket_free( &ticket_ctx );
#endif
#if defined(MBEDTLS_SSL_COOKIE_C)
    mbedtls_ssl_cookie_free( &cookie_ctx );
#endif
```

```
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    const char* message = mbedtls_test_helper_is_psa_leaking();
    if( message )
    {
        if( ret == 0 )
            ret = 1;
        mbedtls_printf( "PSA memory leak detected: %s\n",  message);
    }
#endif /* MBEDTLS_USE_PSA_CRYPTO */
```

```
mbedtls_ssl_free( &ssl );
mbedtls_ssl_config_free( &conf );
rng_free( &rng );
```

>  Can you please add a note in the documentation of `rng_free`

Yes, I will add a note.

Reply:
Ah, great, so there's no library bug then, just the wrong order in the test code. Your comment mostly has the right order: first free SSL features, then free crypto features. By this logic, `mbedtls_test_helper_is_psa_leaking()` and `mbedtls_psa_crypto_free()` should come before `mbedtls_ssl_free()` and `mbedtls_ssl_config_free()`.

Note that the call to `mbedtls_psa_crypto_free` must stay, for configurations with `MBEDTLS_USE_PSA_CRYPTO` but without `MBEDTLS_TEST_USE_PSA_CRYPTO_RNG`.

Reply:
> By this logic, `mbedtls_test_helper_is_psa_leaking()` and `mbedtls_psa_crypto_free()` should come before `mbedtls_ssl_free()` and `mbedtls_ssl_config_free()`.

This seems to contradict the previous sentence, didn't you mean "after"?

Note: there's a pre-existing thing that's not an issue in practice, but not clean in principle, which is that we're free-ing things like certs and PKs which `ssl_config` refers to, before freeing `ssl_config` itself. Also, on the server, the various SSL bits are dispersed. I know this goes slightly beyond the original scope of this PR, but can we use the following order for cleanup in both client and server:

1. Sockets
2. All `mbedtls_ssl_xxx_free()` functions starting with `ssl_free` then `ssl_config_free`.
3. Various buffers held by `main()`: `buf`, `session_data`, `context_buf` + structures specific to the programs like `sni_info` and `psk_info`
4. X.509 and associated PK context and PSA slots
5. Crypto stuff like DHM and remaining PSA slots.
6. The test for PSA leaks
7. Then the call(s) to `mbedtls_psa_crypto_free()` (if applicable) and `rng_free()`.
8. Then the other tests (test hooks, memory buffer allocator)

Reply:
@mpg  Changed the order of clean up as requested.